### DISCUSSION TITLE
The way to start mitmproxy from python script

### DISCUSSION THREAD

**Discussion by Sa1ntVeronika** [mitmproxy/mitmproxy#5255] [Q&A]
Hello!

I am looking for a possibility to start proxy from a python - need to work with requests in my automated tests
I've found solution in https://github.com/mitmproxy/mitmproxy/issues/3306 branch, but it seems that something has changed  - script like below can't start in case of AttributeError

```
from mitmproxy import proxy, options
from mitmproxy.tools.dump import DumpMaster

class AddHeader:
    def request(self, flow):
        if flow.request.pretty_host == "example.org":
            flow.request.host = "mitmproxy.org"

def start():
    myaddon = AddHeader()
    opts = options.Options(listen_host='0.0.0.0', listen_port=8080, mode='transparent', confdir='/home/user/.mitmproxy')
    pconf = proxy.config.ProxyConfig(opts)
    m = DumpMaster(opts)
    m.server = proxy.server.ProxyServer(pconf)
    m.addons.add(myaddon)

    try:
        m.run()
    except KeyboardInterrupt:
        m.shutdown()

start()
```

`
`AttributeError: module 'mitmproxy.proxy' has no attribute 'config'`
`
So is there an alternate solution?
mitmproxy            8.0.0

Thanks for any responses!

**Comment by roniemartinez** [comment]
```python
from mitmproxy.tools.main import mitmdump

mitmdump(args=["-s", "myaddon.py"])
```

**Reply by wangzhatian** [reply]
mitmdump_list = ["-s", "addon{}addons_main.py".format(os.path.sep),
                     "--listen-host", "0.0.0.0",
                     "-p", "{}".format(port),
                     "--ignore-hosts '.*443$'",
                     "--set block_global=false"]

--set block_global=false参数貌似没生效
```
Only processing flows that match "--ignore-hosts '.*443$' & --set block_global=false"
Loading script addon\addons_main.py
Proxy server listening at http://*:55555
168.13.2.70:62767: client connect
Client connection from 168.13.2.70 killed by block_global option.
168.13.2.70:62767: client kill connection
168.13.2.70:62767: client disconnect
168.13.2.70:62768: client connect
```

**Reply by roniemartinez** [reply]
@wangzhatian 

> 参数貌似没生效

> Translated: The parameter doesn't seem to take effect

If this was an issue for you, you probably need to open an issue ticket in https://github.com/mitmproxy/mitmproxy/issues

**Reply by mhils** [reply]
Instead of
```python
                 "--ignore-hosts '.*443$'",
                 "--set block_global=false"]
```
you want 
```python
                 "--ignore-hosts", ".*443$",
                 "--set", "block_global=false"]
```

**Reply by Sa1ntVeronika** [reply]
Okay, I got it, thanks a lot. But is it possible to reach requests in my test? Something like
```
def test_func():
    resp = mitmproxy.method_to_work_with_responce
    my_resp = '{jsonlike string}'
    assert resp == my_resp
    
```

Or is it possible to work with requests/responses only via addons?

**Reply by wangzhatian** [reply]
Thank you. Where can I find relevant documents for this part of content

**Reply by ldcvanderpoel** [reply]
How can the mitmdump process be terminated? Previously this was done with running `shutdown()` on the object returned by`DumpMaster()`. How should this be done when starting mitmdump via `mitmdump(args=["-s", "myaddon.py"])`?

**Comment by m0rt1c** [comment]
I am using a script like this to run the proxy:

```python
#!/bin/env python
import asyncio
import sys
from mitmproxy import options
from mitmproxy.tools import dump

class RequestLogger:
    def request(self, flow):
        print(flow.request)

async def start_proxy(host, port):
    opts = options.Options(listen_host=host, listen_port=port)

    master = dump.DumpMaster(
        opts,
        with_termlog=False,
        with_dumper=False,
    )
    master.addons.add(RequestLogger())
    
    await master.run()
    return master

if __name__ == '__main__':
    host=sys.argv[1]
    port=int(sys.argv[2])
    asyncio.run(start_proxy(host, port))
```
The only tricky part for me was to install mitmproxy in my virtual env from the sources:
```bash
git clone https://github.com/mitmproxy/mitmproxy.git
cd mitmproxy
pip install .
```
Also my python version is `3.10.4`.

**Reply by Sa1ntVeronika** [reply]
thanks a lot. Is it possible to shutdown proxy using your methods? Something like new function with `master.shutdown` inside? Or executing mitm commands?

**Reply by m0rt1c** [reply]
I don't know what is your use case is, but could do something like this to shutdown the proxy:
```python
class RequestLogger:
    def __init__(self, master):
        self.master = master

    def request(self, flow):
        # you can achieve this with a command like `curl -X SHUTDOWN -x host:port host`
        if flow.request.method == "SHUTDOWN":
            self.master.shutdown()
            return
        print(flow.request)
```
And create the addon as:
```python
master.addons.add(RequestLogger(master))
```

**Reply by danifersonaglia** [reply]
Anyone know how to get this: mitmproxy --mode upstream:http://185.72.242.22:5705 but only using python script (inside script)?
to run only like this:
mitmproxy -s script.py

**Comment by danifersonaglia** [comment]
Anyone know how to get this:
mitmproxy --mode upstream:http://185.72.242.22:5705/
but only using python script (inside script)?

to run only like this:
mitmproxy -s script.py

**Reply by saucer-man** [reply]
opts = options.Options(listen_host="127.0.0.1", listen_port=port,  mode=[f'upstream:{proxy}'])

**Comment by saucer-man** [comment]
in cmd,I can`mitmdump -p 9090 --mode upstream:http://127.0.0.1:10800--upstream-auth aaa:bbb`，but in Options, I can not add upstream-auth to mitmproxy.options.Options, It will raise as 
```
KeyError: 'Unknown options: upstream_auth'
```
